protocol-units/da-sequencer/node/src/celestia/mod.rs

```diff
@@ -1,93 +1,160 @@
+		})?;
+		let block = block
+			.ok_or(DaSequencerError::BlockRetrieval(format!("Block at {:?} not found", at)))?;
+		Ok(block)
+	}
+}
+
+impl<C: CelestiaClient + Sync + Clone> DaSequencerExternalDa for CelestiaExternalDa<C> {
+	/// Send the given block to Celestia.
+	/// The block is not immediately sent but aggregated in a blob
+	/// until the client can send it to celestia.
+	async fn send_block(&self, _block: SequencerBlockDigest) -> Result<(), DaSequencerError> {
 		todo!()
 	}
 
 	/// Get the blob from celestia at the given height.
-	pub async fn get_blob_at_height(&self) -> Result<Option<Blob>, DaSequencerError> {
+	async fn get_blobs_at_height(
+		&self,
+		_height: CelestiaHeight,
+	) -> Result<Option<Vec<Blob>>, DaSequencerError> {
 		todo!()
 	}
 
-	/// Bootstrap the Celestia client to recovert from missing block.
+	/// Bootstrap the Celestia client to recover from missing block.
 	/// In case of crash for example, block sent to Celestia can be behind the block created by the network.
-	/// The role of this function is to recovert this missing block to all block of the network are sent to celestia.
-	/// The basic algorythm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
-	/// The missing block. Not sure last_notified_celestia_height is useful.
+	/// The role of this function is to recover this missing block to all block of the network are sent to celestia.
+	/// The basic algorithm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
+	/// the missing block. Not sure last_notified_celestia_height is useful.
 	/// During this boostrap new block are sent to the client.
 	/// These block should be buffered until the boostrap is done then sent after in order.
-	pub async fn bootstrap(
+	async fn bootstrap(
 		&self,
 		current_block_height: BlockHeight,
-		last_sent_block_height: BlockHeight,
-		last_notified_celestia_height: CelestiaHeight,
+		last_finalized_celestia_height: CelestiaHeight,
 	) -> Result<(), DaSequencerError> {
-		todo!()
+		// wait to ensure that no blob is pending in the Celestia network
+		tokio::time::sleep(DELAY_SECONDS_BEFORE_BOOTSTRAPPING).await;
+
+		// Step 1: Get last digest in the last finalized blob
+		let digest = self
+			.get_blobs_at_height(last_finalized_celestia_height)
+			.await?
+			.and_then(|mut blobs| blobs.pop())
+			.and_then(|mut digest| digest.0.pop())
+			.ok_or(DaSequencerError::ExternalDaBootstrap(format!(
+				"Celestia returned no blobs or an empty last blob at height {}",
+				last_finalized_celestia_height.0
+			)))?;
+
+		// Step 2: Get the Block for digest
+		let mut block = self.request_block(BlockAt::Digest(digest.clone())).await?;
+
+		// Step 3: Request and send all missing blocks
+		for height in (block.height.0 + 1)..=current_block_height.0 {
+			block = self.request_block(BlockAt::Height(BlockHeight(height))).await?;
+			self.send_block(block.get_block_digest()).await?;
```

Comment:
During the bootstrap process, new blocks digests can be sent by the main loop using `CelestiaClient.send_block()`.
You should add a way to differentiate these block that must be sent after yours to keep the order of all blocks.

Reply:
Yes you can merge, and you'll do new PRs for the 2 issues (`last_finalized_celestia_height` and `send_block()`)